### PR TITLE
Fix Issue 12118 - Modify immutable data using throw

### PR DESCRIPTION
**Pull request by ntrel**
Deprecate throwing an immutable object.

### CODE REVIEW

**Comment by tim-dlang** [comment]
It is still possible to throw an immutable object by first converting it to const:

```
immutable Dummy e = new immutable Dummy(data);
const Dummy e2 = e;
throw e2;
```

Maybe throwing const objects should be deprecated, too. Or could that break too much code?

**Comment by pbackus** [comment]
In principle, the correct solution here is to require `catch (const Exception e)` or `catch (immutable Exception e)` for catching an `immutable(Exception)`. But even if we plan to implement that eventually, it may be best to go through a deprecation period first to avoid silently changing the behavior of existing code.

In addition to `const` and `immutable`, this should probably also apply to `shared` and `inout`.

**Comment by ntrel** [comment]
> But even if we plan to implement that eventually, it may be best to go through a deprecation period first to avoid silently changing the behavior of existing code.

I have made a pull to deprecate catching immutable: https://github.com/dlang/dmd/pull/14707

> In addition to const and immutable, this should probably also apply to shared and inout.

OK, now done. shared and inout already caused:
> ice10651.d(13): Error: can only throw class objects derived from `Throwable`, not type `int*`

Which was a weird message for `throw new shared Exception`. That now gets a can't throw qualifed type error.

**Comment by ntrel** [comment]
> That now gets a can't throw qualifed type error.

Turns out Phobos uses throw shared:

> std/concurrency.d(2272): Deprecation: cannot throw object of qualified type `shared(Throwable)`

Changed to deprecate any const or immutable type.

**Comment by pbackus** [comment]
Throwing a `shared` object is still problematic, even if Phobos currently does it and gets away with it. For example:

```d
shared int n;

class E : Exception
{
    int* p;
    this(shared(int)* p) shared
    {
        super("");
        this.p = p;
    }
}

shared E e = new shared(E)(&n);

void main()
{
    import std.range, std.parallelism, core.atomic;

    foreach (i; iota(1000).parallel)
    {
        try
        {
            throw atomicLoad(e);
        }
        catch (E e) // casts away shared
        {
            *e.p += 1; // data race
        }
    }

    // almost always fails
    assert(atomicLoad(n) == 1000);
}
```
Maybe this PR is not the place to address `throw shared`, but it does need to be dealt with at some point.

**Comment by ssvb** [comment]
> In principle, the correct solution here is to require `catch (const Exception e)` or `catch (immutable Exception e)` for catching an `immutable(Exception)`.

Would it be possible to automatically infer the const/immutable qualifiers in the catch blocks? I mean, automatically replace
```D
catch (Exception e) { /* do some read-only stuff with 'e' */ }
```
with
```D
catch (immutable Exception e) { /* do some read-only stuff with 'e' */ }
```
and allow it to catch immutable exceptions. While the catch blocks like
```D
catch (Exception e) { /* do some read/write stuff with 'e' */ }
```
wouldn't get the immutable qualifier and wouldn't handle the thrown immutable exceptions. This way only the incorrect code would be broken. I also want to mention that if the `static immutable` objects were actually allocated in the read-only data segment (similar to string literals), then this could have prevented the bugs of this kind since a long time ago.

> But even if we plan to implement that eventually, it may be best to go through a deprecation period first to avoid silently changing the behavior of existing code.

This doesn't look like a good solution, because it kinda kills any prospects for having a `safe` and simplistic `nogc` exceptions handling in the near future (various DIPs can take years before anything usable is implemented). The current situation is that D compilers can do such exceptions handling with a bit of extra care. This pull request actually changes it for the worse.

**Comment by pbackus** [comment]
> Would it be possible to automatically infer the const/immutable qualifiers in the catch blocks?

Perhaps, but only in extremely simple cases. Qualifiers can affect overload resolution, so any function or method call that uses the exception (including calling `toString` to print the stack trace) would cause inference to fail.

Also, IMHO it would be poor language design to infer mutability only for `catch` block parameters and not for other variables. Ideally we would like to resolve this issue without adding new special cases to the language.

> This doesn't look like a good solution, because it kinda kills any prospects for having a safe and simplistic nogc exceptions handling in the near future (various DIPs can take years before anything usable is implemented). The current situation is that D compilers can do such exceptions handling with a bit of extra care. This pull request actually changes it for the worse.

I agree it's not ideal, and I sympathize with your impatience, but rushing ahead with quick-and-dirty solutions instead of taking the time to fix things properly is exactly how we end up with problems like this in the first place.

**Comment by ntrel** [comment]
> In principle, the correct solution here is to require catch (const Exception e) or catch (immutable Exception e) for catching an immutable(Exception).

I think that would break working `nothrow` functions that do `catch (Exception)` to call any function that might throw an exception. Those would need to change to catch `const(Exception)`.

**Comment by RazvanN7** [comment]
Yeah, I think that disallowing the possibility to throw qualified exceptions is the path go. However, it is worrisome that this PR only does it for `const`  and `immutable`. We should just deprecate all qualifiers and fix the phobos. Are there problems with `inout` also?

Also, this would need a spec PR and a changelog entry.

**Comment by pbackus** [comment]
What's the plan for `shared`?

**Comment by ssvb** [comment]
> > This doesn't look like a good solution, because it kinda kills any prospects for having a safe and simplistic nogc exceptions handling in the near future (various DIPs can take years before anything usable is implemented). The current situation is that D compilers can do such exceptions handling with a bit of extra care. This pull request actually changes it for the worse.
> 
> I agree it's not ideal

Actually now I see that this isn't a problem at all. A very simple workaround is just to strip the immutable qualifier by a cast before throwing it. This suppresses the deprecation warning and the old code keeps working just like it did before. This cast can't be carelessly misused from `@safe` code and has to be `@trusted`.

The safety of throwing a `static immutable` exception can be potentially hardened by placing it into a read-only section. So that nobody can covertly write to it without getting caught. I'm trying to see if this can be done: https://forum.dlang.org/post/vvbfulclpvoojxrwhkuq@forum.dlang.org

> and I sympathize with your impatience,

This is a misunderstanding. I'm only interested in preventing the existing code from getting broken by this particular pull request and I'm not impatiently waiting for anything. That said, the https://github.com/dlang/dmd/pull/14710 pull request looks like one of the necessary steps towards eventually making [DIP1008](https://github.com/dlang/DIPs/blob/master/DIPs/other/DIP1008.md) usable and enabled by default.

**Comment by RazvanN7** [comment]
@pbackus I don't think that special casing shared is the way to go. We either accept all qualifiers for throwing/catching or we accept neither. Why would we want to have different behavior?

**Comment by pbackus** [comment]
@RazvanN7 That's exactly why I'm asking—the PR currently special-cases `shared`:

```d
if (!exp.type.isNaked() && exp.type.mod != MODFlags.shared_)
```

**Comment by RazvanN7** [comment]
@ntrel There are outstanding reviews for this PR. Could you maybe explain what is the plan for throwing `shared` exceptions?

**Comment by ntrel** [comment]
@RazvanN7 I'm not sure how to fix Phobos which throws a shared exception. If anyone knows then I'm happy to remove the workaround.

**Comment by RazvanN7** [comment]
@ntrel Can you point out the exact code that is failing?

**Comment by ntrel** [comment]
See https://github.com/dlang/dmd/pull/14706#issuecomment-1355286419. (It was the buildkite Phobos test that was failing).

**Comment by pbackus** [comment]
The code in question is here: https://github.com/dlang/phobos/blob/98de7a5bfc091d857be681e8a69b7719bcc6781e/std/concurrency.d#L2272-L2273

It's not exactly a fix, but the easiest way to get rid of the special case would be to have that code cast away `shared` directly. It's no worse than what the runtime is already doing, and the code is `@system` anyway, since it uses `Variant`.

**Comment by ntrel** [comment]
buildkite failing to build dmd:
```
The following operation failed:
Name: common
```
I don't know why.

**Comment by pbackus** [comment]
I think this is the problem:

> `Command: /buildkite/builds/geod24-main-1/dlang/dmd/buildkite-ci-build/distribution/bin/dmd -c -of/buildkite/builds/geod24-main-1/dlang/dmd/dmd/generated/linux/release/64/common.o -version=MARS -w -de -fPIC -m64 -J/buildkite/builds/geod24-main-1/dlang/dmd/dmd/generated/linux/release/64 -I/buildkite/builds/geod24-main-1/dlang/dmd/dmd/compiler/src -dip25 -g -color=on src/dmd/common/bitfields.d src/dmd/common/file.d src/dmd/common/int128.d src/dmd/common/outbuffer.d src/dmd/common/string.d`
>
> `-----------------------------------------------------------`
>
> ``Deprecation: `-dip25` no longer has any effect``

The build command is using the `-dip25` switch, but that switch is deprecated, and `-de` turns the deprecation into an error.

**Comment by ntrel** [comment]
Rebased, tests now passing. This now deprecates throwing shared objects too.

File: changelog/dmd.throw-qualifier.dd [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+Previously, an `immutable`, `const`, `inout` or `shared` exception could be
+thrown and then caught in an unqualified `catch (Exception e)` clause.
+That breaks type safety.
+Throwing a qualified object is now deprecated. This helps to prevent
+possible mutation of an immutable object in a `catch` clause.
+
+The runtime also modifies a thrown object (e.g. to contain a stack
+trace) which can violate `const` or `immutable` objects. Throwing
+qualified objects have been deprecated for this reason also.
```
Comment (RazvanN7) [inline changelog/dmd.throw-qualifier.dd:11]:
```suggestion
qualified objects has been deprecated for this reason also.
```

**Comment by RazvanN7** [comment]
cc @WalterBright @atilaneves . I think we can go ahead and deprecate this. Then, if need be, we can reintroduce this via a DIP.

**Review by dkorpel — APPROVED** [review]

**Comment by ntrel** [comment]
> I'm less sure about shared. 

The status quo doesn't prevent shared being caught as unqualified: https://github.com/dlang/dmd/pull/14706#issuecomment-1355352525
How else would that be prevented?

> How is collection collecting a thread's exception supposed to work?

Not sure I follow. Catching a shared type will still be allowed (#14707 only deprecates catching an immutable type).

> how are staticError and friends supposed to work?

staticError has a constraint `if (is(T : Error))` so it doesn't allow qualified types. (As it reuses storage it can't support immutable anyway). Which friends?

**Comment by thewilsonator** [comment]
>  Which friends?

I thought there was also a `staticException`, perhaps I was misremembering that.

**Comment by atilaneves** [comment]
> cc @WalterBright @atilaneves . I think we can go ahead and deprecate this. Then, if need be, we can reintroduce this via a DIP.

I agree.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**